File: src/utils/conversionValidator.js
/**
 * Post-Conversion Validation Utility
 *
 * Performs quality checks on converted documents to detect:
 * - Page count discrepancies
 * - Missing content
 * - Font embedding issues
 * - File size anomalies
 * - Text extraction failures
 */

const fs = require('fs');
const path = require('path');
const pdfParse = require('pdf-parse');

class ConversionValidator {
  // Validation thresholds and constants
  static THRESHOLDS = {
    MIN_PDF_SIZE_BYTES: 10000, // Minimum PDF size (10KB) - smaller may indicate conversion failure
    MAX_PDF_SIZE_MULTIPLIER: 10, // PDF should not be more than 10x the source size
    MAX_TEXT_EXTRACTION_RATIO: 0.5, // Maximum ratio of extracted text to file size
  };

  /**
   * Validate a DOCX to PDF conversion
   *
   * @param {string} docxPath - Original DOCX file
   * @param {string} pdfPath - Generated PDF file
   * @returns {Promise<Object>} Validation result
   */
  async validateDocxToPdf(docxPath, pdfPath) {
    const issues = [];
    const warnings = [];
    const info = {};

    try {
      // Check files exist
      if (!fs.existsSync(docxPath)) {
        issues.push('Source DOCX file not found');
        return { valid: false, issues, warnings, info };
      }

      if (!fs.existsSync(pdfPath)) {
        issues.push('Output PDF file not found');
        return { valid: false, issues, warnings, info };
      }

      // Get file sizes
      const docxStats = fs.statSync(docxPath);
      const pdfStats = fs.statSync(pdfPath);

      info.docxSize = (docxStats.size / 1024).toFixed(2) + ' KB';
      info.pdfSize = (pdfStats.size / 1024).toFixed(2) + ' KB';

      // Parse PDF for detailed analysis
      const pdfBuffer = fs.readFileSync(pdfPath);
      const pdfData = await pdfParse(pdfBuffer);

      info.pdfPages = pdfData.numpages;
      info.pdfVersion = pdfData.version || 'Unknown';

      // Check if PDF is suspiciously small
      if (pdfStats.size < this.THRESHOLDS.MIN_PDF_SIZE_BYTES) {
        warnings.push('PDF file is very small - conversion may have failed');
      }

      // Check if PDF is suspiciously large
      if (pdfStats.size > docxStats.size * this.THRESHOLDS.MAX_PDF_SIZE_MULTIPLIER) {
        warnings.push('PDF is significantly larger than source - possible image quality issue');
      }

      // Extract text for content validation
      const extractedText = pdfData.text || '';
      const wordCount = extractedText.trim().split(/\s+/).length;

      info.extractedWordCount = wordCount;

      // Check if document appears empty
      if (wordCount < 10) {
        issues.push('Very little text extracted from PDF - conversion likely failed');
      } else if (wordCount < 50) {
        warnings.push('Low word count detected - verify conversion completeness');
      }

      // Check for common error indicators in text
      const errorPatterns = [
        /error/i,
        /failed/i,
        /exception/i,
        /could not/i,
        /unable to/i
      ];

      const firstPage = extractedText.substring(0, 500);
      for (const pattern of errorPatterns) {
        if (pattern.test(firstPage)) {
          warnings.push('Potential error message detected in PDF content');
          break;
        }
      }

      // Font embedding check (basic)
      // LibreOffice should embed fonts, but we can check file size as proxy
      const sizeRatio = pdfStats.size / docxStats.size;
      if (sizeRatio < 0.1) {
        warnings.push('PDF much smaller than DOCX - fonts may not be properly embedded');
      }

      // Page count estimation (very rough)
      // Average DOCX is ~20KB per page, PDF is ~50KB per page
      const estimatedDocxPages = Math.max(1, Math.round(docxStats.size / 20480));
      const actualPdfPages = pdfData.numpages;

      info.estimatedSourcePages = estimatedDocxPages;
      info.actualPdfPages = actualPdfPages;

      const pageDifference = Math.abs(actualPdfPages - estimatedDocxPages);
      const pageDiscrepancyPercent = (pageDifference / estimatedDocxPages) * 100;

      if (pageDiscrepancyPercent > 20) {
        warnings.push(`Page count may be off - estimated ${estimatedDocxPages} pages, got ${actualPdfPages} pages`);
      }

      // Check for metadata
      if (pdfData.info) {
        info.pdfProducer = pdfData.info.Producer || 'Unknown';
        info.pdfCreator = pdfData.info.Creator || 'Unknown';

        // Verify LibreOffice produced this
        if (info.pdfProducer && !info.pdfProducer.includes('LibreOffice')) {
          warnings.push('PDF not produced by LibreOffice - unexpected conversion method');
        }
      }

      // Determine overall validity
      const valid = issues.length === 0;

      return {
        valid,
        issues,
        warnings,
        info,
        recommendation: this._getRecommendation(valid, issues, warnings)
      };

    } catch (error) {
      issues.push(`Validation error: ${error.message}`);
      return {
        valid: false,
        issues,
        warnings,
        info,
        recommendation: 'Could not complete validation - check PDF file integrity'
      };
    }
  }

  /**
   * Get recommendation based on validation results
   *
   * @param {boolean} valid - Overall validity
   * @param {Array} issues - Critical issues
   * @param {Array} warnings - Non-critical warnings
   * @returns {string} Recommendation
   * @private
   */
  _getRecommendation(valid, issues, warnings) {
    if (issues.length > 0) {
      return 'REJECT: Critical issues detected - conversion failed';
    }

    if (warnings.length === 0) {
      return 'ACCEPT: Conversion appears successful with no warnings';
    }

    if (warnings.length <= 2) {
      return 'REVIEW: Minor warnings detected - manual review recommended';
    }

    return 'CAUTION: Multiple warnings detected - verify output quality';
  }

  /**
   * Compare two PDFs (original vs converted)
   *
   * @param {string} expectedPdfPath - Original/expected PDF
   * @param {string} actualPdfPath - Converted PDF to validate
   * @returns {Promise<Object>} Comparison result
   */
  async comparePdfs(expectedPdfPath, actualPdfPath) {
    const differences = [];
    const info = {};

    try {
      if (!fs.existsSync(expectedPdfPath)) {
        throw new Error('Expected PDF not found');
      }

      if (!fs.existsSync(actualPdfPath)) {
        throw new Error('Actual PDF not found');
      }

      // Parse both PDFs
      const expectedBuffer = fs.readFileSync(expectedPdfPath);
      const actualBuffer = fs.readFileSync(actualPdfPath);

      const expectedData = await pdfParse(expectedBuffer);
      const actualData = await pdfParse(actualBuffer);

      info.expectedPages = expectedData.numpages;
      info.actualPages = actualData.numpages;

      // Page count comparison
      if (expectedData.numpages !== actualData.numpages) {
        const diff = actualData.numpages - expectedData.numpages;
        differences.push({
          type: 'page_count',
          severity: 'high',
          message: `Page count mismatch: expected ${expectedData.numpages}, got ${actualData.numpages} (${diff > 0 ? '+' : ''}${diff})`
        });
      }

      // Text extraction comparison (basic)
      const expectedText = expectedData.text || '';
      const actualText = actualData.text || '';

      const expectedWords = expectedText.trim().split(/\s+/).length;
      const actualWords = actualText.trim().split(/\s+/).length;

      info.expectedWords = expectedWords;
      info.actualWords = actualWords;

      const wordDiff = actualWords - expectedWords;
      const wordDiffPercent = Math.abs((wordDiff / expectedWords) * 100);

      if (wordDiffPercent > 5) {
        differences.push({
          type: 'word_count',
          severity: 'medium',
          message: `Word count differs by ${wordDiffPercent.toFixed(1)}%: expected ~${expectedWords}, got ~${actualWords}`
        });
      }

      // File size comparison
      const expectedSize = fs.statSync(expectedPdfPath).size;
      const actualSize = fs.statSync(actualPdfPath).size;

      info.expectedSize = (expectedSize / 1024).toFixed(2) + ' KB';
      info.actualSize = (actualSize / 1024).toFixed(2) + ' KB';

      const sizeDiffPercent = Math.abs(((actualSize - expectedSize) / expectedSize) * 100);

      if (sizeDiffPercent > 50) {
        differences.push({
          type: 'file_size',
          severity: 'low',
          message: `File size differs significantly: expected ${info.expectedSize}, got ${info.actualSize}`
        });
      }

      // Calculate similarity score (0-100)
      const pageScore = expectedData.numpages === actualData.numpages ? 100 : 0;
      const wordScore = Math.max(0, 100 - wordDiffPercent);
      const overallScore = (pageScore * 0.7 + wordScore * 0.3);

      info.similarityScore = overallScore.toFixed(1);

      return {
        matching: differences.length === 0,
        differences,
        info,
        summary: this._getComparisonSummary(overallScore, differences)
      };

    } catch (error) {
      return {
        matching: false,
        differences: [{ type: 'error', severity: 'critical', message: error.message }],
        info,
        summary: 'Comparison failed'
      };
    }
  }

  /**
   * Get comparison summary
   *
   * @param {number} score - Similarity score (0-100)
   * @param {Array} differences - List of differences
   * @returns {string} Summary text
   * @private
   */
  _getComparisonSummary(score, differences) {
    if (score >= 95) {
      return 'Excellent match - conversion is highly accurate';
    } else if (score >= 85) {
      return 'Good match - minor differences acceptable';
    } else if (score >= 70) {
      return 'Fair match - noticeable differences present';
    } else {
      return 'Poor match - significant differences detected';
    }
  }

  /**
   * Format validation report for console output
   *
   * @param {Object} result - Validation result
   * @returns {string} Formatted report
   */
  formatReport(result) {
    let report = '\n';
    report += '╔══════════════════════════════════════════════════════════╗\n';
    report += '║   Conversion Validation Report                           ║\n';
    report += '╚══════════════════════════════════════════════════════════╝\n\n';

    // Status
    const status = result.valid ? ' VALID' : ' INVALID';
    report += `Status: ${status}\n`;
    report += `Recommendation: ${result.recommendation}\n\n`;

    // Info
    if (Object.keys(result.info).length > 0) {
      report += 'Document Info:\n';
      for (const [key, value] of Object.entries(result.info)) {
        report += `  • ${key}: ${value}\n`;
      }
      report += '\n';
    }

    // Issues
    if (result.issues.length > 0) {
      report += ' Critical Issues:\n';
      result.issues.forEach(issue => {
        report += `  • ${issue}\n`;
      });
      report += '\n';
    }

    // Warnings
    if (result.warnings.length > 0) {
      report += '  Warnings:\n';
      result.warnings.forEach(warning => {
        report += `  • ${warning}\n`;
      });
      report += '\n';
    }

    if (result.issues.length === 0 && result.warnings.length === 0) {
      report += ' No issues detected\n\n';
    }

    return report;
  }
}

module.exports = new ConversionValidator();

File: FileConversionApi/Utils/ConversionValidator.cs
using Microsoft.Extensions.Logging;
using iText.Kernel.Pdf.Canvas.Parser;
using iText.Kernel.Pdf;
using System.Diagnostics;

namespace FileConversionApi.Utils;

/// <summary>
/// Post-conversion validation utility
/// Performs quality checks on converted documents to detect issues
/// </summary>
public class ConversionValidator : IConversionValidator
{
    private readonly ILogger<ConversionValidator> _logger;

    // Validation thresholds and constants
    private const long MinPdfSizeBytes = 10000; // Minimum PDF size (10KB) - smaller may indicate conversion failure
    private const double MaxPdfSizeMultiplier = 10.0; // PDF should not be more than 10x the source size
    private const double MaxTextExtractionRatio = 0.5; // Maximum ratio of extracted text to file size

    public ConversionValidator(ILogger<ConversionValidator> logger)
    {
        _logger = logger;
    }

    /// <inheritdoc/>
    public async Task<ConversionValidationResult> ValidateDocxToPdfAsync(string docxPath, string pdfPath)
    {
        var result = new ConversionValidationResult
        {
            InputPath = docxPath,
            OutputPath = pdfPath,
            ConversionType = "DOCX->PDF",
            Issues = new List<string>(),
            Warnings = new List<string>(),
            Info = new Dictionary<string, object>()
        };

        try
        {
            // Check files exist
            if (!File.Exists(docxPath))
            {
                result.Issues.Add("Source DOCX file not found");
                result.IsValid = false;
                return result;
            }

            if (!File.Exists(pdfPath))
            {
                result.Issues.Add("Output PDF file not found");
                result.IsValid = false;
                return result;
            }

            // Get file sizes
            var docxInfo = new FileInfo(docxPath);
            var pdfInfo = new FileInfo(pdfPath);
            var docxSize = docxInfo.Length;
            var pdfSize = pdfInfo.Length;

            result.Info["docxSize"] = docxSize;
            result.Info["pdfSize"] = pdfSize;

            // Validate PDF size
            if (pdfSize < MinPdfSizeBytes)
            {
                result.Issues.Add($"PDF file too small ({pdfSize} bytes) - possible conversion failure");
                result.IsValid = false;
            }

            if (pdfSize > docxSize * MaxPdfSizeMultiplier)
            {
                result.Warnings.Add($"PDF file unusually large ({pdfSize} vs {docxSize} bytes source)");
            }

            // Extract and validate text content
            var textExtraction = await ExtractPdfTextAsync(pdfPath);
            if (textExtraction.Success)
            {
                result.Info["extractedTextLength"] = textExtraction.Text.Length;
                result.Info["pageCount"] = textExtraction.PageCount;

                // Check for empty PDF
                if (textExtraction.Text.Length == 0 && textExtraction.PageCount == 0)
                {
                    result.Issues.Add("PDF appears to be empty - no text or pages detected");
                    result.IsValid = false;
                }

                // Check text extraction ratio
                if (textExtraction.Text.Length > 0)
                {
                    var textRatio = (double)textExtraction.Text.Length / pdfSize;
                    if (textRatio > MaxTextExtractionRatio)
                    {
                        result.Warnings.Add($"Unusually high text extraction ratio ({textRatio:P2})");
                    }
                }
            }
            else
            {
                result.Warnings.Add($"Text extraction failed: {textExtraction.Error}");
            }

            // Validate file integrity
            if (!await IsValidPdfAsync(pdfPath))
            {
                result.Issues.Add("PDF file appears to be corrupted or invalid");
                result.IsValid = false;
            }

            result.IsValid = result.Issues.Count == 0;

            _logger.LogInformation("DOCX->PDF validation completed. Valid: {Valid}, Issues: {Issues}, Warnings: {Warnings}",
                result.IsValid, result.Issues.Count, result.Warnings.Count);

        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "DOCX->PDF validation failed");
            result.Issues.Add($"Validation error: {ex.Message}");
            result.IsValid = false;
        }

        return result;
    }

    /// <inheritdoc/>
    public async Task<ConversionValidationResult> ValidateXlsxToCsvAsync(string xlsxPath, string csvPath)
    {
        var result = new ConversionValidationResult
        {
            InputPath = xlsxPath,
            OutputPath = csvPath,
            ConversionType = "XLSX->CSV",
            Issues = new List<string>(),
            Warnings = new List<string>(),
            Info = new Dictionary<string, object>()
        };

        try
        {
            // Check files exist
            if (!File.Exists(xlsxPath))
            {
                result.Issues.Add("Source XLSX file not found");
                result.IsValid = false;
                return result;
            }

            if (!File.Exists(csvPath))
            {
                result.Issues.Add("Output CSV file not found");
                result.IsValid = false;
                return result;
            }

            // Get file sizes
            var xlsxInfo = new FileInfo(xlsxPath);
            var csvInfo = new FileInfo(csvPath);

            result.Info["xlsxSize"] = xlsxInfo.Length;
            result.Info["csvSize"] = csvInfo.Length;

            // Validate CSV content
            var csvContent = await File.ReadAllTextAsync(csvPath);
            if (string.IsNullOrWhiteSpace(csvContent))
            {
                result.Issues.Add("CSV file is empty");
                result.IsValid = false;
            }
            else
            {
                // Count lines and check for basic CSV structure
                var lines = csvContent.Split('\n', StringSplitOptions.RemoveEmptyEntries);
                result.Info["csvLines"] = lines.Length;

                if (lines.Length == 0)
                {
                    result.Issues.Add("CSV file contains no data lines");
                    result.IsValid = false;
                }
            }

            result.IsValid = result.Issues.Count == 0;

        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "XLSX->CSV validation failed");
            result.Issues.Add($"Validation error: {ex.Message}");
            result.IsValid = false;
        }

        return result;
    }

    /// <inheritdoc/>
    public async Task<ConversionValidationResult> ValidateCsvToXlsxAsync(string csvPath, string xlsxPath)
    {
        var result = new ConversionValidationResult
        {
            InputPath = csvPath,
            OutputPath = xlsxPath,
            ConversionType = "CSV->XLSX",
            Issues = new List<string>(),
            Warnings = new List<string>(),
            Info = new Dictionary<string, object>()
        };

        try
        {
            // Check files exist
            if (!File.Exists(csvPath))
            {
                result.Issues.Add("Source CSV file not found");
                result.IsValid = false;
                return result;
            }

            if (!File.Exists(xlsxPath))
            {
                result.Issues.Add("Output XLSX file not found");
                result.IsValid = false;
                return result;
            }

            // Get file sizes
            var csvInfo = new FileInfo(csvPath);
            var xlsxInfo = new FileInfo(xlsxPath);

            result.Info["csvSize"] = csvInfo.Length;
            result.Info["xlsxSize"] = xlsxInfo.Length;

            // Basic XLSX validation (we can't easily read XLSX content without additional libraries)
            if (xlsxInfo.Length < 4096) // XLSX files are typically at least 4KB
            {
                result.Issues.Add("XLSX file unusually small - possible conversion failure");
                result.IsValid = false;
            }

            result.IsValid = result.Issues.Count == 0;

        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "CSV->XLSX validation failed");
            result.Issues.Add($"Validation error: {ex.Message}");
            result.IsValid = false;
        }

        return result;
    }

    /// <inheritdoc/>
    public async Task<ConversionValidationResult> ValidateImageConversionAsync(string inputPath, string outputPath, string conversionType)
    {
        var result = new ConversionValidationResult
        {
            InputPath = inputPath,
            OutputPath = outputPath,
            ConversionType = conversionType,
            Issues = new List<string>(),
            Warnings = new List<string>(),
            Info = new Dictionary<string, object>()
        };

        try
        {
            // Check files exist
            if (!File.Exists(inputPath))
            {
                result.Issues.Add("Source image file not found");
                result.IsValid = false;
                return result;
            }

            if (!File.Exists(outputPath))
            {
                result.Issues.Add("Output image file not found");
                result.IsValid = false;
                return result;
            }

            // Get file sizes
            var inputInfo = new FileInfo(inputPath);
            var outputInfo = new FileInfo(outputPath);

            result.Info["inputSize"] = inputInfo.Length;
            result.Info["outputSize"] = outputInfo.Length;

            // Check if output is reasonably sized
            if (outputInfo.Length < 1024) // Images should be at least 1KB
            {
                result.Issues.Add("Output image file too small - possible conversion failure");
                result.IsValid = false;
            }

            // Check if output is not excessively large (more than 50x input)
            if (outputInfo.Length > inputInfo.Length * 50)
            {
                result.Warnings.Add("Output image significantly larger than input");
            }

            result.IsValid = result.Issues.Count == 0;

        }
        catch (Exception ex)
        {
            _logger.LogError(ex, "Image conversion validation failed");
            result.Issues.Add($"Validation error: {ex.Message}");
            result.IsValid = false;
        }

        return result;
    }

    /// <summary>
    /// Extract text from PDF for validation
    /// </summary>
    private async Task<PdfTextExtractionResult> ExtractPdfTextAsync(string pdfPath)
    {
        var result = new PdfTextExtractionResult();

        try
        {
            await using var pdfStream = new FileStream(pdfPath, FileMode.Open, FileAccess.Read);
            using var pdfDoc = new PdfDocument(new PdfReader(pdfStream));

            result.PageCount = pdfDoc.GetNumberOfPages();
            var text = new System.Text.StringBuilder();

            for (int i = 1; i <= result.PageCount; i++)
            {
                var page = pdfDoc.GetPage(i);
                var pageText = PdfTextExtractor.GetTextFromPage(page);
                text.Append(pageText);
            }

            result.Text = text.ToString();
            result.Success = true;
        }
        catch (Exception ex)
        {
            result.Success = false;
            result.Error = ex.Message;
        }

        return result;
    }

    /// <summary>
    /// Basic PDF validity check
    /// </summary>
    private async Task<bool> IsValidPdfAsync(string pdfPath)
    {
        try
        {
            await using var pdfStream = new FileStream(pdfPath, FileMode.Open, FileAccess.Read);
            using var pdfDoc = new PdfDocument(new PdfReader(pdfStream));
            return pdfDoc.GetNumberOfPages() > 0;
        }
        catch
        {
            return false;
        }
    }
}

/// <summary>
/// PDF text extraction result
/// </summary>
internal class PdfTextExtractionResult
{
    public bool Success { get; set; }
    public string Text { get; set; } = string.Empty;
    public int PageCount { get; set; }
    public string? Error { get; set; }
}

/// <summary>
/// Conversion validation result
/// </summary>
public class ConversionValidationResult
{
    public bool IsValid { get; set; } = true;
    public string? InputPath { get; set; }
    public string? OutputPath { get; set; }
    public string? ConversionType { get; set; }
    public List<string> Issues { get; set; } = new();
    public List<string> Warnings { get; set; } = new();
    public Dictionary<string, object> Info { get; set; } = new();
}

/// <summary>
/// Conversion validator interface
/// </summary>
public interface IConversionValidator
{
    Task<ConversionValidationResult> ValidateDocxToPdfAsync(string docxPath, string pdfPath);
    Task<ConversionValidationResult> ValidateXlsxToCsvAsync(string xlsxPath, string csvPath);
    Task<ConversionValidationResult> ValidateCsvToXlsxAsync(string csvPath, string xlsxPath);
    Task<ConversionValidationResult> ValidateImageConversionAsync(string inputPath, string outputPath, string conversionType);
}
